libsolidity/formal/CHC.cpp

```diff
@@ -1101,186 +1085,182 @@ pair<CheckResult, CHCSolverInterface::CexGraph> CHC::query(smtutil::Expression c
+				verificationTargets.push_back(CHCVerificationTarget{
+					{target.type, placeholder.fromPredicate, placeholder.constraints && placeholder.errorExpression == target.errorId },
+					target.errorId,
+					target.errorNode
+				});
 			}
-		}
 	}
-}
 
-void CHC::checkAssertTarget(ASTNode const* _scope, CHCVerificationTarget const& _target)
-{
-	solAssert(_target.type == VerificationTarget::Type::Assert, "");
-	auto assertions = transactionAssertions(_scope);
-	for (auto const* assertion: assertions)
+	set<unsigned> checkedErrorIds;
+	for (auto const& target: verificationTargets)
 	{
-		auto it = m_errorIds.find(assertion->id());
-		solAssert(it != m_errorIds.end(), "");
-		solAssert(!it->second.empty(), "");
-		unsigned errorId = it->second[0];
+		string satMsg;
+		string satMsgUnderflow;
+		string satMsgOverflow;
+		string unknownMsg;
+		ErrorId errorReporterId;
+		ErrorId underflowErrorId = 3944_error;
+		ErrorId overflowErrorId = 4984_error;
+
+		if (target.type == VerificationTarget::Type::PopEmptyArray)
+		{
+			solAssert(dynamic_cast<FunctionCall const*>(target.errorNode), "");
+			satMsg = "Empty array \"pop\" detected here.";
+			unknownMsg = "Empty array \"pop\" might happen here.";
+			errorReporterId = 2529_error;
+		}
+		else if (
+			target.type == VerificationTarget::Type::Underflow ||
+			target.type == VerificationTarget::Type::Overflow
+		)
+		{
+			auto const* expr = dynamic_cast<Expression const*>(target.errorNode);
+			solAssert(expr, "");
+			auto const* intType = dynamic_cast<IntegerType const*>(expr->annotation().type);
+			if (!intType)
+				intType = TypeProvider::uint256();
+
+			satMsgUnderflow = "Underflow (resulting value less than " + formatNumberReadable(intType->minValue()) + ")";
+			satMsgOverflow = "Overflow (resulting value larger than " + formatNumberReadable(intType->maxValue()) + ")";
+			if (target.type == VerificationTarget::Type::Underflow)
+			{
+				satMsg = satMsgUnderflow + " happens here.";
+				unknownMsg = satMsgUnderflow + " might happen here.";
+				errorReporterId = underflowErrorId;
+			}
+			else if (target.type == VerificationTarget::Type::Overflow)
+			{
+				satMsg = satMsgOverflow + " happens here.";
+				unknownMsg = satMsgOverflow + " might happen here.";
+				errorReporterId = overflowErrorId;
+			}
+		}
+		else if (target.type == VerificationTarget::Type::DivByZero)
+		{
+			satMsg = "Division by zero happens here.";
+			unknownMsg = "Division by zero might happen here.";
+			errorReporterId = 4281_error;
+		}
+		else if (target.type == VerificationTarget::Type::Assert)
+		{
+			satMsg = "Assertion violation happens here.";
+			unknownMsg = "Assertion violation might happen here.";
+			errorReporterId = 6328_error;
+		}
+		else
+			solAssert(false, "");
 
-		checkAndReportTarget(assertion, _target, errorId, 6328_error, "Assertion violation happens here.", "Assertion violation might happen here.");
+		checkAndReportTarget(target, errorReporterId, satMsg, unknownMsg);
```

Comment:
```suggestion
		checkAndReportTarget(target, errorReporterId, errorType + " happens here.", errorType + " might happen here.");
```
We could also remove `satMsg` and `unknownMsg` and have a single `errorType` that has, for example,
"Array empty \"pop\""
"Underflow ...."
"Overflow ...."
"Division by zero"
etc
which is filled inside each branch above, and then use this line instead

Reply:
Updated according to the suggestion. The only thing is that the message for the `pop` target changed from `detected here` to `happens here`, so I had to update the tests accordingly.